src/ConcatenateMsg.chpl

```diff
@@ -130,14 +184,24 @@ module ConcatenateMsg
                         start = 0;
                         for (name, i) in zip(names, 1..) {
                             // lookup and cast operand to copy from
-                            var o = toSymEntry(st.lookup(name), int);
-                            ref ea = e.a;
-                            // copy array into concatenation array
-                            forall (i, v) in zip(o.aD, o.a) with (var agg = newDstAggregator(int)) {
-                              agg.copy(ea[start+i], v);
+                            const o = toSymEntry(st.lookup(name), int);
+                            if mode == "interleave" {
+                              coforall loc in Locales {
+                                on loc {
+                                  const size = o.aD.localSubdomain().size;
+                                  e.a[blockstarts[here.id]..#size] = o.a.localSlice[o.aD.localSubdomain()];
```

Comment:
@ronawho @bradcray I'd love your opinion on this line, from a comms perspective. The source is local, and the destination is either entirely local or it might have one non-local element on the end(s), depending on how the BlockDist boundaries line up. Is a range slice on the destination the best pattern to use, or should I be using a domain slice? Or something else?

Reply:
Tagging @e-kayrakli as well, who's been looking at slicing recently and may have better data than I do.

While I'm thinking about it, just to check on terminology, when you say the destination is local, do you mean "non-distributed" or "located on the same locale as the source" (modulo those end conditions you mention?).

Reply:
Good question -- I mean located on the same locale as the source (i.e. `loc` from the `on` block), modulo end conditions.

Reply:
If there's a chance that the slice can expand more than one locale, then keep in mind that you cannot just slice the array by `{blockstarts[here.id]..#size}` as the slice won't inherit the distribution of the domain. So, you'll first need to slice the domain of that array and then pass that domain slice (e.g. `myArr[myArr.domain[start..end]]` and not `myArr[{start..end}]`)

I can't think of any comm-related gotchas between the two. However, if you do domain-based slicing, I think what will happen is that we'll recognize the newly created domain as constant, but we may not recognize that it is a "single-use" domain. So, its creation will involve local and non-contentious locking (to count its arrays). Whereas, if you do range-based slicing, the array code will create a domain internally, and it'll be aware that it is a single-use one. So it will avoid that lock. Obviously this matters only if your data movement is small.

Reply:
@e-kayrakli Thanks for the info! Just to confirm, the code as-written (using a range-based slice) will create a single-use, internal domain that inherits the distribution of the destination array's domain... do I have that right?

Reply:
Engin got back here before me, but I'm not sure I agree with what he says.  :)  I believe it's legal to slice a distributed array with a non-distributed domain and that the differences relate more to what the domain of the resulting expression will be rather than whether one is legal or not (e.g., if I said `var A = myDistArr[lo..hi];`, A would be a distributed array itself, even if it's only distributed to a single locale.  Whereas if I said `var A = myDistArr[{lo..hi}];`, then A would be a local array, definitely on this locale.  It shouldn't matter whether the indices `lo..hi` are local to the current locale for the original array or not.

Though I felt confident in the previous statement, I wasn't actually certain which would be cheaper for this pattern, so did some quick experiments and was going to advise writing the LHS as `e.a[{blockstarts[here.id]..#size}]` because I think it will be cheaper.  Specifically, slicing a distributed array with a range currently creates a new distributed domain, which is a fairly expensive operation involving all locales; whereas slicing it with a domain does not.

This is regrettable since the slice-by-range syntax is cleaner, and we're looking into ways to reduce those overheads.  Using the `-schpl_serializeSlices=true` flag at compile-time helps (for either idiom), but I recall we've talked about using this before with Arkouda, but aren't in the Makefile today, so maybe there were other downsides that I'm not remembering.

Here's the test I was using in making these characterizations:

```chapel
use BlockDist, CommDiagnostics;

var A = newBlockArr(1..1000, real),
    B = newBlockArr(1..1000, real);

resetCommDiagnostics();
startCommDiagnostics();
A[1..10] = B.localSlice[11..20];
stopCommDiagnostics();
printCommDiagnosticsTable();

resetCommDiagnostics();
startCommDiagnostics();
A[{1..10}] = B.localSlice[11..20];
stopCommDiagnostics();
printCommDiagnosticsTable();


resetCommDiagnostics();
startCommDiagnostics();
A[{492..501}] = B.localSlice[11..20];
stopCommDiagnostics();
printCommDiagnosticsTable();


resetCommDiagnostics();
startCommDiagnostics();
A[492..501] = B.localSlice[11..20];
stopCommDiagnostics();
printCommDiagnosticsTable();
```

Reply:
I stand corrected. I wasn't implying that it would be illegal to slice it with a local domain, but I was expecting that it would incur some fine grained communication (though reading what I wrote, I haven't made it clear). But looking a bit more, it looks like it is not the case and I agree with Brad that you should slice with a domain.

@reuster986 said

> Just to confirm, the code as-written (using a range-based slice) will create a single-use, internal domain that inherits the distribution of the destination array's domain... do I have that right?

That's right, yes. It'll in fact slice the array's domain to create the new one. However, it'll do some more internal adjustments to make that domain a bit lighter-weight compared to a regular user-level domain.

Reply:
Thank you, both! Sorry, I spent all of yesterday dealing with a strange python 3.9 issue and did not get around to incorporating your feedback until today, but I just switched to domain-based slicing. I'll start a separate issue to discuss using  `-schpl_serializeSlices=true`.